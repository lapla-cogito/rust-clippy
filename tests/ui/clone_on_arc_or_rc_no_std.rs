#![no_std]
#![warn(clippy::clone_on_arc_or_rc)]

extern crate alloc;
use alloc::rc::Rc;
use alloc::string::String;
use alloc::sync::Arc;
use alloc::vec;
use alloc::vec::Vec;

fn no_std() {
    let arc: Arc<String> = Arc::new("foo".into());
    let _: String = (*arc).clone();
    //~^ error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone

    let rc: Rc<String> = Rc::new("foo".into());
    let _: String = (*rc).clone();
    //~^ error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone

    let arc: Arc<String> = Arc::new("foo bar".into());
    let _: Vec<_> = (*arc).clone().split(" ").collect();
    //~^ error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone

    let rc: Rc<Vec<u32>> = Rc::new(vec![1, 2, 3]);
    let _: Vec<_> = (*rc).clone().iter().map(|x| x + 1).collect();
    //~^ error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone

    let _: String = (*Arc::<String>::new("foo".into())).clone();
    //~^ error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone
}
