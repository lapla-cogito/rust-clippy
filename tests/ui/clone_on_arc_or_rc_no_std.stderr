error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone
  --> tests/ui/clone_on_arc_or_rc_no_std.rs:13:21
   |
LL |     let _: String = (*arc).clone();
   |                     ^^^^^^^^^^^^^^ help: try: `alloc::sync::Arc::unwrap_or_clone(arc)`
   |
   = note: `-D clippy::clone-on-arc-or-rc` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(clippy::clone_on_arc_or_rc)]`

error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone
  --> tests/ui/clone_on_arc_or_rc_no_std.rs:17:21
   |
LL |     let _: String = (*rc).clone();
   |                     ^^^^^^^^^^^^^ help: try: `alloc::rc::Rc::unwrap_or_clone(rc)`

error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone
  --> tests/ui/clone_on_arc_or_rc_no_std.rs:21:21
   |
LL |     let _: Vec<_> = (*arc).clone().split(" ").collect();
   |                     ^^^^^^^^^^^^^^ help: try: `alloc::sync::Arc::unwrap_or_clone(arc)`

error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone
  --> tests/ui/clone_on_arc_or_rc_no_std.rs:25:21
   |
LL |     let _: Vec<_> = (*rc).clone().iter().map(|x| x + 1).collect();
   |                     ^^^^^^^^^^^^^ help: try: `alloc::rc::Rc::unwrap_or_clone(rc)`

error: conditionally unwrapping an `Arc`/`Rc` may avoid unnecessary clone
  --> tests/ui/clone_on_arc_or_rc_no_std.rs:28:21
   |
LL |     let _: String = (*Arc::<String>::new("foo".into())).clone();
   |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `alloc::sync::Arc::unwrap_or_clone(Arc::<String>::new("foo".into()))`

error: aborting due to 5 previous errors

